### ISSUE TITLE
storeChunk: shared_ptr only?

### ISSUE THREAD

**Issue by ax3l** [openPMD/openPMD-api#140]
Quick question regarding the write `storeChunk` API.
Why does it only accept `shared_ptr`s as parameters?

I was playing around with use cases where user data is allocated in containers such as `std::array` and `std::vector` in which the data layout is guaranteed to be [contiguous and access is well-defined](http://en.cppreference.com/w/cpp/container/vector) ([ContiguousContainer](http://en.cppreference.com/w/cpp/concept/ContiguousContainer)s), e.g. via `vector<t>.data()`. How do we get such containers' data written? :-)

```C++
// user code
std::vector<int> v = {1, 1, 3, 4, 7, 11};

// write me, I am famous
```

Do we need a [`ContiguousIterator`](http://en.cppreference.com/w/cpp/concept/ContiguousIterator) interface as well? (Dang, that concept exists in `std::vector` since C++03 but the `ContiguousIterator` is C++17... Anyway, there are just three std lib types where this is relevant for now so we could overload or define traits. Fun fact: `vector<bool>` is special and not guaranteed to be contiguous.)

Anyway, the data could come from any user-data as well which might be badly designed C pointer. We can't enforce this backwards on the user to change their containers/data storage/etc. just to write. How do we handle such data? If we just add a `unique_ptr<T>` interface this would mean we would have to move this ptr out again when we are done writing it. Verbose and cumbersome...

Saying that, another question regarding the existing `shared_ptr` interface: when should the shared ptr reference be released by the API again? Imho, first then it is save to use the ptr at user-side again, so after `.close()` and `.flush()`.

**Comment by anokfireball** [comment]
The main reason for employing ```shared_ptr```s was to make the shared ownership during unfinished IO explicit. Raw C-style pointers were also considered, but discarded as they give no insight into what happens to them over time.

ContiguousContainers store data in exactly the way ```storeChunk``` expects - as one unstrided contiguous slab of values, one after the other. So to get the values of a vector
```c++
std::vector< int > x = /* ... */;
```
you would just put the ```std::vector::data()```-pointer into a ```shared_ptr```. To not have double frees from both the vector and the```shared_ptr```, the deleter needs to be empty. You end up with
```c++
std::shared_ptr< int > x_chunk{
    x.data(),
    [](int*){  }
};
```
which you can use to write your data as was intended
```c++
Offset x_offset = /* ... */;
Extent x_extent = { x.size() };
// x_chunk.use_count() == 1
series.iteration[1].particles["e"]["position"]["x"].storeChunk(x_offset, x_extent, x_chunk);
// x_chunk.use_count() >= 2 before and during flush
series.flush();
// x_chunk.use_count() == 1
```

This addresses your last question as well:
The reference to the ```shared_ptr``` is released precisely when the ```IOTask``` that contains it (an ```Operation::WRITE_DATASET```) gets popped off the the IO queue.
As we do not have fine-grained control of the queue exposed to the user (*yet*), your assumption/opinion is entirely correct: The user may reuse the resource as soon as the ```Series::flush()``` with the respective ```shared_ptr``` has completed.

**Comment by ax3l** [comment]
> The main reason for employing `shared_ptrs` was to make the shared ownership during unfinished IO explicit.

Yep, I really like the incentive.

>  The reference to the shared_ptr is released precisely when the IOTask that contains it (an Operation::WRITE_DATASET) gets popped off the the IO queue.

Excellent, nice! :sparkles: That was my implied question.

> just put the `std::vector::data()`-pointer into a `shared_ptr`

Yes, that's what I thought & did initially as well. Unfortunately, a `std::shared_ptr<int>( v.data() )` will always lead to a double free when the shared pointer is destroyed (because a vector can not *share* the access to its data).

> the deleter needs to be empty

neat trick, thanks!

Nevertheless, I don't think that's good design if the user input is already rather "raw". `std::vector` is fully unaware its data is shared, so why add the burden and verbosity. It does not add to any safety at all in such a case and is just boilerplate. Or not? I am open to discussion on this, it's just a bit confusing imho.

**Comment by ax3l** [comment]
Implementing your empty-deleter recommendation above I am thinking we could otherwise also keep the single-interface and add a util/helper of the form "raw pointer passing" or "unsave passing" for the cases mentioned above (user-given containers without proper sharing ability):

```C++
namespace openPMD
{
    template<typename T>
    std::shared_ptr<T>
    passUnsave( T* x )
    {
        return std::shared_ptr<T>(x, [](T*){});
    }
} // openPMD

E.storeChunk(offset, extent, passUnsave(v.data()));
```

What do you think?

**Comment by anokfireball** [comment]
The intention behind what is now [```auxiliary::allocatePtr()```](https://github.com/openPMD/openPMD-api/blob/dev/include/openPMD/auxiliary/Memory.hpp#L15) was something along these lines (the user-side API changed and it became unneccessary, is now used internally).

Personally, I'd like to keep one single interface for load/store (unless we introduce xtensor at some point). It just keeps the interface more clean rather than overloading it with more types.
The core of your idea should probalby be implemented much like you propsed. I'd prefer a different name, to make it explicit that it is there to *temporarily* share ownership of the raw data that is passed to the funciton. Additionally, if we decide to offer such functionality, we should include overloads for some of the C++ STL types you proposed:

```c++
namespace openPMD
{
    template<typename T>
    std::shared_ptr<T>
    shareOwnership( T* x );

    template<typename T>
    std::shared_ptr<T>
    shareOwnership( std::vector< T > );

    template<typename T, std::size_t T_size>
    std::shared_ptr<T>
    shareOwnership( std::array< T, size > );
} // openPMD
```

**Comment by ax3l** [comment]
Sounds good, maybe `shareRaw` would be even more specific as a name?

**Comment by ax3l** [comment]
@C0nsultant huh, what are we going to do with the `unique_ptr< T[] > &` of the `loadChunk` API and raw pointers/containers?
Added a draft for it in #204 as well.

**Comment by anokfireball** [comment]
Closed in #204.
